grails-app/domain/org/pih/warehouse/inventory/CycleCount.groovy

```diff
@@ -23,10 +25,35 @@ class CycleCount {
     def beforeInsert() {
         createdBy = AuthService.currentUser
         updatedBy = AuthService.currentUser
+        status = recomputeStatus()
     }
 
     def beforeUpdate() {
         updatedBy = AuthService.currentUser
+        status = recomputeStatus()
+    }
+
+    static hasMany = [
+            /*
+             * We've moved away from using "hasMany" in our code whenever possible due to performance concerns around
+             * inserting new child records (the whole set of items will always be fetched when addToCycleCountItems is
+             * called). In this specific case we've deemed it okay because:
+             *
+             * 1) cycleCountItems is a relatively small dataset (almost always < 100 elements, usually < 10) and it is
+             *    not ever-growing because a cycle count doesn't stay open forever.
+             *
+             * 2) When we're working with a cycle count, we almost always want to fetch all of its items anyways for
+             *    validation purposes.
+             *
+             * Using hasMany greatly simplifies recomputing the cycle count status since it allows the cycle count to
+             * maintain an accurate list of items (which it uses to compute its status). Still, the disadvantage is
+             * important to note, and we should strive to replace this hasMany with a better option in the future.
+             */
+            cycleCountItems: CycleCountItem,
+    ]
+
+    static mapping = {
+        cycleCountItems cascade: 'all-delete-orphan'
```

Comment:
I would like to have a convention around the cascading. There are pros and cons of having/not having cascade behavior. Due to the cascading we have sometimes run into unexpected issues difficult to spot. It's better to have some of the operations being done explicitly (in this case deleting).

Reply:
yeah I agree. We don't ever delete cycle counts so I can change this to only cascade updates if we want, but I figured if we ever did need to delete a cycle count, we'd also want to delete the items (otherwise we'll get constraint violations)